stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -313,13 +319,83 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
+                    dataArray.getType().getTag() == TypeTags.UNION_TAG);
+        } else {
+            if (fieldTypeTag == TypeTags.UNION_TAG) {
+                bStruct.setRefField(refRegIndex.incrementAndGet(), null);
+            } else {
+                handleNilToNonNillableFieldAssignment();
+            }
+        }
+    }
+
+    private void handleMappingArrayValue(BType nonNilType, BStruct bStruct, BNewArray dataArray,
+            RegistryIndex refRegIndex, boolean containsNull) {
+        if (nonNilType.getTag() == TypeTags.ARRAY_TAG) {
+            BArrayType expectedArrayType = (BArrayType) nonNilType;
+            if (((BArrayType) nonNilType).getElementType().getTag() == TypeTags.UNION_TAG) {
+                handleMappingArrayElementToUnionType(expectedArrayType, dataArray, refRegIndex, bStruct);
+            } else {
+                handleMappingArrayElementToNonUnionType(containsNull, nonNilType, dataArray, bStruct, refRegIndex);
+            }
         } else {
-            bStruct.setRefField(refRegIndex.incrementAndGet(), getDataArray(dataArray));
+            handleMismatchingFieldAssignment();
+        }
+    }
+
+    private void handleMappingArrayElementToNonUnionType(boolean containsNull, BType nonNilType, BNewArray newArray,
+            BStruct bStruct, RegistryIndex refRegIndex) {
+        if (containsNull) {
+            throw new BallerinaException(
+                    "Trying to assign an array containing NULL values to an array of a non-nillable " + "element type");
+        } else {
+            validateAndSetNonUnionRefRecordField(bStruct, refRegIndex.incrementAndGet(), ((BArrayType) nonNilType)
+                    .getElementType().getTag(), ((BArrayType) newArray.getType()).getElementType().getTag(), newArray);
+        }
+    }
+
+    private void handleMappingArrayElementToUnionType(BArrayType expectedArrayType, BNewArray arrayTobeSet,
+            RegistryIndex refRegIndex, BStruct bStruct) {
+        BType actualArrayElementType = arrayTobeSet.getType();
+        if (actualArrayElementType.getTag() == TypeTags.NULL_TAG) {
+            bStruct.setRefField(refRegIndex.incrementAndGet(), arrayTobeSet);
+        } else {
+            BUnionType expectedArrayElementUnionType = (BUnionType) expectedArrayType.getElementType();
+            BType expectedNonNilArrayElementType = retrieveNonNilType(expectedArrayElementUnionType.getMemberTypes());
+            BType actualNonNilArrayElementType = getActualNonNilArrayElementType(actualArrayElementType);
+            validateAndSetUnionRefRecordField(bStruct, refRegIndex.incrementAndGet(), expectedNonNilArrayElementType
+                    .getTag(), actualNonNilArrayElementType.getTag(), arrayTobeSet);
+        }
+    }
+
+    private BType getActualNonNilArrayElementType(BType actualArrayElementType) {
+        if (actualArrayElementType.getTag() == TypeTags.UNION_TAG) {
+            return (((BUnionType) actualArrayElementType).getMemberTypes()).get(0);
```

Comment:
Can this be always 0? for ex: int|() and ()|int both valid are valid

Reply:
The element is created inside the code[1] . In the code we are adding to the list in that order, so it is always 0.

[1] https://github.com/ballerina-platform/ballerina-lang/pull/8588/files#diff-7b71bf65c3c83159943e19af40ab847dR356